vllm/entrypoints/openai/tool_parsers.py

```diff
@@ -0,0 +1,624 @@
+                        # try parsing it with regular JSON - if it works we're at the end, and we need to send the
+                        #   difference between tokens streamed so far and the valid JSON
+                        delta = None
+
+                # check to see if the name is defined and has been sent. if so, stream the name - otherwise keep waiting
+                # finish by setting old and returning None as base case
+                self.prev_tool_call_arr = tool_call_arr
+                return delta
+
+            except Exception as e:
+                logger.error(f'Error trying to handle streaming tool call: {e}')
+                logger.info(f'Skipping chunk as a result of tool streaming extraction error')
+                return None
+
+
+class Hermes2ProToolParser(ToolParser):
+    tool_call_start_token: str = '<tool_call>'
+    tool_call_end_token: str = '</tool_call>'
+    tool_call_start_token_id: int = 128004
+    tool_call_end_token_id: int = 128011
+
+    # regex to match between <tool_call> and </tool_call> OR between <tool_call> and EOS (happens sometimes :))
+    tool_call_regex = re.compile(r'<tool_call>(.*?)</tool_call>|<tool_call>(.*)', re.DOTALL)
+    scratch_pad_regex = re.compile(r'<scratch_pad>(.*?)</scratch_pad>', re.DOTALL)
+
+    @staticmethod
+    def extract_tool_calls(model_output: str) -> ExtractedToolCallInformation:
+
+        # sanity check; avoid unnecessary processing
+        if Hermes2ProToolParser.tool_call_start_token not in model_output:
+            return ExtractedToolCallInformation(
+                tools_called=False,
+                tool_calls=[],
+                content=model_output
+            )
+
+        else:
+
+            try:
+                # there are two possible captures - between tags, or between a tag and end-of-string so the result of findall
+                #   is an array of tuples where one is a function call and the other is None
+                function_call_tuples = Hermes2ProToolParser.tool_call_regex.findall(model_output)
+
+                # load the JSON, and then use it to build the Function and Tool Call
+                raw_function_calls = [json.loads(match[0] if match[0] else match[1]) for match in function_call_tuples]
+                tool_calls = [
+                    ToolCall(
+                        type='function',
+                        function=FunctionCall(
+                            name=function_call['name'],
+                            # function call args are JSON but as a string
+                            arguments=json.dumps(function_call['arguments'])
+                        )
+                    ) for function_call in raw_function_calls
+                ]
+
+                # TODO extract including the scratch pad into content
+                content_match = Hermes2ProToolParser.scratch_pad_regex.search(model_output)
+                content = content_match.group(1) if content_match else None
```

Comment:
# Issue

I'm not sure if I misunderstand anything, but the behavior of `extract_tool_calls` and `extract_tool_calls_streaming` seems to differ when the model outputs the following text:
```
I need to call X because Y.
<tool_call>
...
</tool_call>
```

`extract_tool_calls` will produce `message = {'content': None, 'tool_calls': ...}` due to L429 and L430.

On the other hand, `extract_tool_calls_streaming` will produce `message = {'content': 'I need to call X because Y.', 'tool_calls': ...}`

# Suggestion

Maybe we can replace
```py
content_match = Hermes2ProToolParser.scratch_pad_regex.search(model_output)
content = content_match.group(1) if content_match else None
```
with
```py
content = model_output[:model_output.find(Hermes2ProToolParser.tool_call_start_token)]
```
so that `extract_tool_calls` and `extract_tool_calls_streaming` produce the same result?

By the way, it seems that the chatml template jinja doesn't mention `<scratch_pad>` anyway. Since scratch_pad is a pain to deal with in the streaming mode, maybe we don't need scratch_pad in the parser after all?

Reply:
we're currently training Hermes-3 with `<scratch_pad>` which will be returned as part of the content in addition to tool calls when it is part of the jinja template system prompt

Reply:
Great point about streaming handling this differently from non-streaming. made this change. 
Thanks for catching! I even had a TODO in the code I just forgot about it lol
f7f15fae